crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -135,29 +135,82 @@ pub fn run_enter_schedule<S: States>(world: &mut World) {
+#[derive(SystemSet, Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub enum StateTransitionSet {
+    /// All [`OnEnter`] schedules are executed in this set.
+    OnEnter,
+    /// All [`OnTransition`] schedules are executed in this set.
+    OnTransition,
+    /// All [`OnExit`] schedules are executed in this set.
+    OnExit,
+}
+
+/// Run all state transition systems for the given state type.
+///
+/// This function is deprecated. It is not used internally by Bevy,
+/// and will be removed in a future version.
+#[deprecated(since = "0.13.0")]
+pub fn apply_state_transition<S: States>(world: &mut World) {
+    apply_state_transition_on_enter::<S>(world);
+    apply_state_transition_on_transition::<S>(world);
+    apply_state_transition_on_exit::<S>(world);
+}
+
+/// If a new state is queued in [`NextState<S>`], this system
+/// runs the [`OnExit(exited_state)`] schedule, if it exists.
+pub fn apply_state_transition_on_exit<S: States>(world: &mut World) {
+    let next_state_resource = world.resource::<NextState<S>>();
+    if let Some(entered) = &next_state_resource.0 {
+        let state_resource = world.resource::<State<S>>();
+        if *state_resource != *entered {
+            let exited = state_resource.0.clone();
+            // Try to run the schedules if they exist.
+            world.try_run_schedule(OnExit(exited.clone())).ok();
+        }
+    }
+}
+
+/// If a new state is queued in [`NextState<S>`], this system
+/// runs the [`OnTransition { from: exited_state, to: entered_state }`](OnTransition), if it exists.
+pub fn apply_state_transition_on_transition<S: States>(world: &mut World) {
+    let next_state_resource = world.resource::<NextState<S>>();
+    if let Some(entered) = &next_state_resource.0 {
+        let state_resource = world.resource::<State<S>>();
+        if *state_resource != *entered {
+            let exited = state_resource.0.clone();
+            // Try to run the schedules if they exist.
+            world
+                .try_run_schedule(OnTransition {
+                    from: exited,
+                    to: entered.clone(),
+                })
+                .ok();
+        }
+    }
+}
+
 /// If a new state is queued in [`NextState<S>`], this system:
 /// - Takes the new state value from [`NextState<S>`] and updates [`State<S>`].
-/// - Runs the [`OnExit(exited_state)`] schedule, if it exists.
-/// - Runs the [`OnTransition { from: exited_state, to: entered_state }`](OnTransition), if it exists.
 /// - Runs the [`OnEnter(entered_state)`] schedule, if it exists.
-pub fn apply_state_transition<S: States>(world: &mut World) {
+pub fn apply_state_transition_on_enter<S: States>(world: &mut World) {
     // We want to take the `NextState` resource,
     // but only mark it as changed if it wasn't empty.
+    // Note we update the `State` and `NextState` resources only in on enter,
+    // because previous two systems need both the old and new state.
```

Comment:
This is actually a breaking change in behavior. Previously, `apply_state_transition` would `std::mem::take` the value from `NextState`, then replace the value of `State`. Then it would call `OnExit`, `OnTransition`, `OnEnter`. All in that order. I think it makes sense to preserve this order of operations.

One way of preserving the old ordering is to store the old value from a state transition in a `PreviousState` resource, which would later be used to call `OnExit` and `OnTransition`.

```rust
fn apply_state_transition<S: States>(
    mut next_state: ResMut<NestState<S>>,
    mut state: ResMut<State<S>>,
    mut prev_state: ResMut<PreviousState<S>>,
) {
     // Swappy swap. `next_state` -> `state` -> `prev_state`.
     // Nothing else happens here.
     // `OnEnter` and friends get called elsewhere.
}
```

The `.add_systems` would look something like

```rust
.add_systems((
    apply_state_transition,
    // These SystemSets could be private.
    run_exit_schedule.in_set(RunOnEnter),
    run_transition_schedule.in_set(RunOnTransition),
    run_enter_schedule.in_set(RunOnEnter),
).chain())
.configure_sets((
    RunOnExit,
    RunOnTransition,
    RunOnEnter,
).chain());
```

Reply:
> apply_state_transition

```
fn apply_state_transition<S: States>(
    mut next_state: ResMut<NestState<S>>,
    mut state: ResMut<State<S>>,
    mut prev_state: ResMut<PreviousState<S>>,
) {
     // Swappy swap. `next_state` -> `state` -> `prev_state`.
     // Nothing else happens here.
     // `OnEnter` and friends get called elsewhere.
}
```

to make this work we should make this this function private or rename it to avoid break backwards compatibility too hard — if someone uses this function directly.

> store the old value from a state transition in a PreviousState

This can be done.

Except maybe this "swappy swap" should happen after `OnExit`, because while we are running `OnExit` we are logically still in the previous state.

So do you think it is important to preserve current behavior?

And more generally, is it important to guarantee specific values of `State` and `NextState` resources during state transition?

Reply:
> Except maybe this "swappy swap" should happen after OnExit, because while we are running OnExit we are logically still in the previous state.

I think that if we're running `OnExit(A)` when we leave the state `A`, that means we should already be in the next state.

Reply:
Restored previous behavior:
* `PrevState` now holds previous value
* `State` is updated to new state, and `NextState` is reset to `None` before all three systems run